src/Features/Core/Completion/AbstractCompletionService.cs

```diff
@@ -200,22 +286,52 @@ private class ProviderGroup
             return result;
         }
 
-        private static Task<CompletionItemGroup> GetGroupAsync(
-            ICompletionProvider provider,
+        private static async Task<CompletionList> GetCompletionListAsync(
+            CompletionListProvider provider,
             Document documentOpt,
             SourceText text,
             int position,
             CompletionTriggerInfo triggerInfo,
             CancellationToken cancellationToken)
         {
-            return provider is ITextCompletionProvider
-                ? Task.FromResult(((ITextCompletionProvider)provider).GetGroup(text, position, triggerInfo, cancellationToken))
-                : documentOpt != null
-                    ? provider.GetGroupAsync(documentOpt, position, triggerInfo, cancellationToken)
-                    : SpecializedTasks.Default<CompletionItemGroup>();
+            if (provider is TextCompletionProvider)
+            {
+                return ((TextCompletionProvider)provider).GetCompletionList(text, position, triggerInfo, cancellationToken);
+            }
+
+            if (documentOpt != null)
+            {
+                var itemsBuilder = ImmutableArray.CreateBuilder<CompletionItem>();
+                Action<CompletionItem> addCompletionItem = item =>
+                {
+                    itemsBuilder.Add(item);
+                };
+
+                CompletionItem builder = null;
+                Action<CompletionItem> registerBuilder = item =>
+                {
+                    builder = item;
+                };
+
+                var isExclusive = false;
+                Action<bool> makeExclusive = value =>
+                {
+                    isExclusive = value;
+                };
+
+                var context = new CompletionListContext(documentOpt, position, triggerInfo, addCompletionItem, registerBuilder, makeExclusive, cancellationToken);
+
+                await provider.RegisterCompletionListAsync(context).ConfigureAwait(false);
```

Comment:
I'm matching our existing pattern. See DiagnosticAnalyzer, CodeFixProvider and CodeRefactoringProvider. This is at least consistent.